crimson: add MonClient

tchaikov opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

tchaikov commented:
this is a (buggy) preliminary of MonClient, my plan is to improve it when moving forward. and i am happy to address any concern or comment regarding to it.

tchaikov commented on src/mon/MonMap.cc:
```diff
@@ -438,7 +438,7 @@ int MonMap::init_with_monmap(const std::string& monmap, std::ostream& errout)
 }
 
-int MonMap::init_with_config_file(const ConfigProxy& conf,
+int MonMap::init_with_config_file(const ConfigReader& conf,
```
could avoid this change if we want to specialize the compilation unit for seastar anyway -- we could conditionalize the `#include` the appropriate `config_proxy.h` on `WITH_SEASTAR` anyway.

tchaikov commented on src/common/dout.h:
```diff
@@ -69,6 +76,31 @@ struct is_dynamic<dynamic_marker_t<T>> : public std::true_type {};
+#ifdef WITH_SEASTAR
+#define dout_impl(cct, sub, v)						\
+  do {									\
+    if (ceph::common::local_conf()->subsys.should_gather(sub, v)) {	\
```
could optimize it in the same way used by `WITH_SEASTAR` branch.

tchaikov commented on src/crimson/mon/MonClient.cc:
```diff
@@ -0,0 +1,477 @@
+                        uint32_t want_keys)
+{
+  std::unique_ptr<AuthClientHandler> auth;
+  auth.reset(AuthClientHandler::create(nullptr,
```
`AuthClientHandler` queries config, for instance, for the name of the keyring, etc. so we need a way to fix the calls like `cct->_conf->name` in `CephxClientHandler.cc` .

tchaikov commented:
i am still testing the cephx authorization support.

tchaikov commented:
some of the `#include` cleanups are extracted out into #23930 .

tchaikov commented:
a subset of this PR is extracted out into #23957 in hope to ease the review.

jdurgin commented on src/crimson/mon/MonClient.cc:
```diff
@@ -0,0 +1,500 @@
+    mons.push_back(rank);
+  } else {
+    // const auto parallel =
+    //   ceph::common::local_conf().get_val<uint64_t>("mon_client_hunt_parallel");
```
planning to add this later?

tchaikov replied:
oh, i forgot to uncomment this change when pushing this change. hunting-in-parallel indeed works. will uncomment it once #23957 is merged.

jdurgin commented:
The crimson monclient code looks nice and straight-forward. Seems like a good start. I was trying to see if there were any races possible between future executions within the same function, but didn't find any.

liu-chunmei commented:
@tchaikov except for #23957, do you have other subset PRs?

tchaikov commented:
@liu-chunmei not yet. i will try to get #23957 merged first. and split the changes under `src/auth` directory into another PR.

tchaikov commented:
retest this please.

tchaikov commented:
```
In file included from /opt/rh/devtoolset-7/root/usr/include/c++/7/string:52:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/include/seastar/json/json_elements.hh:24,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/include/seastar/http/api_docs.hh:23,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/src/http/api_docs.cc:22:
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/basic_string.h: In instantiation of 'std::basic_string<_CharT, _Traits, _Alloc>::_If_sv<_Tp, std::basic_string<_CharT, _Traits, _Alloc>&> std::basic_string<_CharT, _Traits, _Alloc>::operator=(_Tp) [with _Tp = seastar::basic_sstring<char, unsigned int, 15>; _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::basic_string<_CharT, _Traits, _Alloc>::_If_sv<_Tp, std::basic_string<_CharT, _Traits, _Alloc>&> = std::basic_string<char>&]':
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/include/seastar/json/json_elements.hh:113:16:   required from 'seastar::json::json_element<T>& seastar::json::json_element<T>::operator=(const C&) [with C = seastar::basic_sstring<char, unsigned int, 15>; T = std::basic_string<char>]'
/home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/include/seastar/http/api_docs.hh:134:27:   required from here
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/basic_string.h:3567:28: error: call of overloaded 'assign(seastar::basic_sstring<char, unsigned int, 15>&)' is ambiguous
  { return this->assign(__sv); }
                            ^
In file included from /opt/rh/devtoolset-7/root/usr/include/c++/7/string:53:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/include/seastar/json/json_elements.hh:24,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/include/seastar/http/api_docs.hh:23,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/src/http/api_docs.cc:22:
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/basic_string.tcc:692:5: note: candidate: std::basic_string<_CharT, _Traits, _Alloc>& std::basic_string<_CharT, _Traits, _Alloc>::assign(const std::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
     basic_string<_CharT, _Traits, _Alloc>::
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/rh/devtoolset-7/root/usr/include/c++/7/string:52:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/include/seastar/json/json_elements.hh:24,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/include/seastar/http/api_docs.hh:23,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests-arm64/src/seastar/src/http/api_docs.cc:22:
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/basic_string.h:4133:7: note: candidate: std::basic_string<_CharT, _Traits, _Alloc>& std::basic_string<_CharT, _Traits, _Alloc>::assign(std::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       assign(basic_string&& __str)
       ^~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/basic_string.h:4232:7: note: candidate: std::basic_string<_CharT, _Traits, _Alloc>& std::basic_string<_CharT, _Traits, _Alloc>::assign(std::basic_string<_CharT, _Traits, _Alloc>::__sv_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::basic_string<_CharT, _Traits, _Alloc>::__sv_type = std::basic_string_view<char>]
       assign(__sv_type __sv)
       ^~~~~~
make[3]: *** [src/seastar/CMakeFiles/seastar.dir/src/http/api_docs.cc.o] Error 1
```

needs to address the FTBFS on aarch64.

tchaikov commented on src/crimson/mon/MonClient.cc:
```diff
@@ -0,0 +1,499 @@
+          });
+        });
+    default:
+      assert(0);
```
yes! as `ceph_assert*()` relies on the logging facility, and when i was working on `ceph::mon::Client`, i didn't put any thoughts on this at first. the logging was added later on when i found it's actually indispensable. will replace it with `ceph_assert_always()`.

tchaikov commented on src/crimson/mon/MonClient.cc:
```diff
@@ -0,0 +1,499 @@
+seastar::future<> Connection::close()
+{
+  if (!closed && conn) {
+    return conn->close();
```
thanks, that's neat!

tchaikov commented on src/crimson/mon/MonClient.cc:
```diff
@@ -0,0 +1,499 @@
+  // we only care about these message types
+  switch (m->get_type()) {
+  case CEPH_MSG_MON_MAP:
+    return handle_monmap(conn, boost::static_pointer_cast<MMonMap>(m));
```
yeah, i feel the same. it's intuitive for the callee to return a future this way. sure, let's fix `SocketMessenger::dispatch()` in a follow-up PR.

cbodley replied:
see https://github.com/ceph/ceph/pull/24090

tchaikov commented:
changelog

- all comments addressed
- added a helper method `make_message()` for creating `Ref<Message>`. because i think the atomic reference counter in `RefCountedObject` is not suitable in seastar, we could just use `boost::intrusive_ref_counter<Message, boost::thread_unsafe_counter>` for seastar-osd. so i am reluctant to reuse the types defined by `MessageSubType`, but i have not figured out the appropriate route to go. so instead of generalizing `MessageSubType` or `RefCountedObject`, i am moving the `Ref` out of the `Message` types in hope that it'd be easier to change latter on.

i still need to:

- https://github.com/ceph/ceph/pull/23849#issuecomment-420685323

tchaikov commented:
retest this please.

tchaikov commented on src/crimson/mon/MonClient.cc:
```diff
@@ -0,0 +1,507 @@
+                                [peer](auto& mc) {
+                                  return mc.is_my_peer(peer);
+                                });
+      assert(found != pending_conns.end());
```
might want to use `ceph_assert_*()`?

tchaikov commented:
changelog

- s/assert()/ceph_assert_always()/
- add a commit to disable libhwloc for silencing the ASan warning on memory leak.

cbodley commented on src/crimson/net/Fwd.h:
```diff
@@ -26,6 +26,14 @@ using auth_proto_t = int;
+struct MMonCommandAck;
+struct MLogAck;
+struct MConfig;
+
```
could these live in MonClient.h or a separate mon/Fwd.h to keep mon details out of crimson/net?

tchaikov replied:
agreed. they don't belong to `net`. will move them into `MonClient.h` .

tchaikov replied:
@cbodley updated and repushed.

tchaikov commented on src/crimson/mon/MonClient.h:
```diff
@@ -0,0 +1,138 @@
+  seastar::timer<seastar::lowres_clock> timer;
+  seastar::gate tick_gate;
+
+  std::shared_ptr<ceph::net::Messenger> msgr;
```
yeah, makes sense!

tchaikov replied:
@cbodley in the latest change, `mon::Client` keeps a reference to `Messenger`. as it should always a valid reference of it. does this sound good to you?

tchaikov commented:
the FTBFS is caused by `operator string_view ()` and `operator std::basic_string<char_type>()` offered by `seastar::sstring` . in C++17, some method and operators also offers `string_view` friendly overloads.  hence the ambiguous:

```
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/basic_string.h:3946:7: note: candidate: std::basic_string<_CharT, _Traits, _Alloc>& std::basic_string<_CharT, _Traits, _Alloc>::operator+=(const std::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]
       operator+=(const basic_string& __str)
       ^~~~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/basic_string.h:3988:7: note: candidate: std::basic_string<_CharT, _Traits, _Alloc>& std::basic_string<_CharT, _Traits, _Alloc>::operator+=(std::basic_string<_CharT, _Traits, _Alloc>::__sv_type) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>; std::basic_string<_CharT, _Traits, _Alloc>::__sv_type = std::basic_string_view<char>]
       operator+=(__sv_type __sv)
       ^~~~~~~~
```

will fix this on seastar side.

tchaikov commented:
#24130 is posted to make sure we have a decent GCC. probably we should do the same in `d/control`.

tchaikov commented:
changelog

- let `mon::Client` hold a reference of `Messenger`
- pass `ceph::net::Messenger&` instead of `ceph::net::SocketMessenger&` in `test_monc.cc`, as what the `MonClient` needs is but a `ceph::net::Messenger`.

cbodley commented on src/test/crimson/test_monc.cc:
```diff
@@ -0,0 +1,72 @@
+        }).then([&monc] {
+          return monc.authenticate(std::chrono::seconds{10});
+        }).finally([&monc] {
+          return monc.stop();
```
i don't see the test calling `msgr.shutdown()`, and i wonder about potential races between that and `monc.stop()`. the messenger still has a pointer to `mon::Client` (via `Dispatcher*`) and may continue making calls to that interface. maybe those ms_ calls in `mon::Client` should have an early return if its been closed?

tchaikov replied:
@cbodley will add `msgr.shutdown()`. regarding to the racing i thought your latest gate change should address it, but it cannot. can we make the `pending_dispatch` a member variable of `SocketConnection` ? so we can wait for it in`SocketConnection::close()` ?

cbodley replied:
that would help with calls to `ms_dispatch()` on connections closed by `mon::Client::close()` - but the messenger could still be making other calls like `ms_accept()` or `ms_connect()`, and it wouldn't help there

i think `mon::Client` itself would need to gate its outstanding futures with `mon::Client::close()`. and in general, any Dispatcher would need to do the same. how about something like a `GatedDispatcher` to encapsulate this and present a new gated interface? ie
```c++
class GatedDispatcher : public Dispatcher {
  seastar::gate gate;
 protected: // present a new protected interface to be overridden
  virtual seastar::future<> gated_dispatch(ConnectionRef conn, MessageRef m) = 0;
  ...

  seastar::future<> gated_close() { return gate.close(); }

 public: // implement public Dispatcher interface in terms of the gated interfaces
  seastar::future<> ms_dispatch(ConnectionRef conn, MessageRef m) override final {
    return seastar::with_gate(gate, [=] { return gated_dispatch(conn, m); });
  }
  ...
};
```
so `mon::Client` would inherit from `GatedDispatcher`, implement the `gated_` interfaces, and call `gated_close()` on `mon::Client::close()`. it's kind of messy, but at least it keeps that mess out of both the Messenger and mon::Client?

tchaikov replied:
@cbodley thanks for the suggestions! i need to ponder on this and then get back to you.

tchaikov replied:
@cbodley i think the major concern here is that the the dispatcher captured by `SocketMessenger` for dispatching the `ms_handle_connect()` and `ms_handle_accept()` calls could be dangling reference when the continuations are scheduled by reactor.

but, in our use case, we attach a single dispatcher to one or more messengers, and this dispatcher instance will outlive the messenger. and we already have a `gate` guarding `SocketMessenger`. so, to avoid adding yet another `with_*` around the inner dispatch call, can we enforce a rule:
 the dispatcher attached to a messenger should be alive until `Messenger::shutdown()` returns, in other words, the messenger will be able to call `Dispatcher`'s methods until `Messenger::shutdown()` returns.

in the case of `mon::Client::stop()`, what it does is just draining all the floating continuations, for instance, the ones scheduled by the timer, and close the connection to monitor. the `mon::Client` instance should still be able to serve as a qualified `Dispatcher`, in the sense that it won't abort or segfault when being notified with calls like `ms_dispatch()` or `ms_handle_reset()`. but i think we can tolerate with exceptions in this case.

what do you think?

cbodley replied:
> in the case of `mon::Client::stop()`, what it does is just draining all the floating continuations, for instance, the ones scheduled by the timer, and close the connection to monitor. the `mon::Client` instance should still be able to serve as a qualified `Dispatcher`, in the sense that it won't abort or segfault when being notified with calls like `ms_dispatch()` or `ms_handle_reset()`. but i think we can tolerate with exceptions in this case.

that all sounds reasonable, yeah - but isn't that basically what the seastar::gate gives us? a point after which new calls will throw an exception?

tchaikov replied:
yeah, `gated_*()`, `gated_close()` and `seastar::gate` ensure us that there is no floating continuations in the wild after that point in a unified way. but we don't get this for free. aside from an extra layer of `Dispatcher` in the code, we need to

1. increment the count in `gate`, throw if it's closed
1. move-copy the continuation and its arguments before applying it
1. attach a continuation to decrement the counter

what i want is to avoid adding another `seastar::with_gate()` in the call stack, if we plant this `gate` before dispatch calls, then each call of `gated_*` method will need to go through the `gate`. and the reason we do this is just for the graceful shutdown, which only happens once in a the whole life of crimson-osd process. i 

currently, `SocketMessenger` already guards all dispatcher calls with `pending_dispatch`, so we can be sure that all `ms_*` calls will be concluded after `SocketMessenger::shutdown()`. if we can also ensure that all `ms_*` calls are still safe *after* `mon::Client::stop()` as you suggested in the [first comment](https://github.com/ceph/ceph/pull/23849#discussion_r218470423) in this conversation thread

> may continue making calls to that interface. maybe those ms_ calls in `mon::Client` should have an early return if its been closed

my argument of not adding `GatedDispatcher` is that i assume the `ms_*` calls are well-behaved  even after `mon::Client::stop()`. and they are allowed to throw exceptions if they are not happy with a closed connection being passed in.

the early return for `ms_*` calls is not able to guarantee that we will be referencing a "stopped" dispatcher in the "later" part of `ms_*` continuation chain after it survives the early check in `ms_*` call for the "closed" status of dispatcher. we could check the "closed" status in every step of `ms_*` call, but that'd be more annoying.  i'd trade it for the grand and unified `GatedDispatcher`.

if you think the well-behaved `ms_*` calls after `stop()` is too strong a requirement for implementing `Dispatcher`s, i am happy to harden it with `seastar::gate`. and yeah, in that case, `GatedDispatcher` is a very good compromise.

cyx1231st replied:
> my argument of not adding GatedDispatcher is that i assume the ms_* calls are well-behaved even after mon::Client::stop(). and they are allowed to throw exceptions if they are not happy with a closed connection being passed in.

IMO there's no need to worry about a closed connection being passed in, if the `dispatch_gate` of the msgr becomes a member of `SocketConnection`. (This is same idea with [the second comment](https://github.com/ceph/ceph/pull/23849/files#r218827807). I just implemented the per-conn `dispatch_gate` in #24142, it is because only the `Protocol` class is calling `Dispatcher` interfaces, and I think it is reasonable...)

> the early return for ms_* calls is not able to guarantee that we will be referencing a "stopped" dispatcher in the "later" part of ms_* continuation chain after it survives the early check in ms_* call for the "closed" status of dispatcher. we could check the "closed" status in every step of ms_* call, but that'd be more annoying. i'd trade it for the grand and unified GatedDispatcher.

Yes, it is also reasonable to me to stop dispatching if the `Dispatcher` is stopped.

> if you think the well-behaved ms_* calls after stop() is too strong a requirement for implementing Dispatchers, i am happy to harden it with seastar::gate. and yeah, in that case, GatedDispatcher is a very good compromise.

It is also possible to instead track the number of registered msgrs of this `Dispatcher`. The `Messenger` need to unregister its `Dispatchers` before shutdown, then the `Dispatcher` just need to wait for the number of msgrs drop to 0 before `stop()` returned. This way, there will be minor performance impact I think.

tchaikov commented:
changelog

- rebased on master
- updated `MonClient.h` to adapt `ceph_mutex.h` changes
- add a commit to expose `-DWITH_SEASTAR=1`
- amend the commit message of the commit which updates the seastar submodule. because back when i started working on this PR, there was a change to refactor the cmake of seastar pending on review, and i wanted to be prepared for its merge, so i reworked our cmake script to adapt the not-yet-merged change. but it turns out this change is not merged yet at the time of writing. and i am facing a linking issue if i revert to the upstream's master. probably we can wait a while before looking at the linking issue.

tchaikov commented:
@cbodley ping? we can always fix tear-down behavior of dispatcher in a follow-up PR as it is orthogonal to this changeset to some degree, IMHO. what do you think?

cbodley commented on src/crimson/mon/MonClient.h:
```diff
@@ -0,0 +1,135 @@
+
+namespace ceph::mon {
+
+class Connection {
```
it looks like Client only needs a forward delcaration of class Connection; consider moving the class itself to MonClient.cc

tchaikov replied:
yeah, this is tempting idea. but `pending_conns` is a `std::vector<Connection>`. i will try to make it a `std::vector<std::unique_ptr<Connection>>` in a follow-up PR.

cbodley replied:
vector should work with forward declarations; you probably just need to define `~Client()` in the source file for access to `~Connection()`
